File: examples/knight/knight.cpp
#include "mxvk/argz.hpp"
#include "mxvk/mxvk.hpp"
#include "mxvk/mxvk_controller.hpp"
#include "mxvk/mxvk_exception.hpp"
#include "mxvk/mxvk_png.hpp"

#include <SDL3/SDL.h>

#include <algorithm>
#include <array>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <format>
#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

namespace knight {
    class Tour {
      public:
        Tour();

        void drawBoard(mxvk::VK_Sprite &whiteCell, mxvk::VK_Sprite &redCell, mxvk::VK_Sprite &visitedCell, float scaleX, float scaleY) const;
        void drawKnight(mxvk::VK_Sprite &texture, float scaleX, float scaleY) const;
        void nextMove();
        void resetTour();
        void resetTour(int startRow, int startCol);
        void resetTourFromPoint(float x, float y);

        [[nodiscard]] int getMoves() const { return moves; }
        [[nodiscard]] bool isTourOver() const { return tourOver; }

      private:
        struct Position {
            int row;
            int col;

            constexpr Position(int newRow = 0, int newCol = 0) : row(newRow), col(newCol) {}
        };

        void initializeBoard();
        void clearBoard();
        [[nodiscard]] bool isValidMove(const Position &position) const;
        [[nodiscard]] int getDegree(const Position &position) const;
        bool solveKnightsTour(Position position, int moveCount);

        static constexpr int BOARD_SIZE = 8;
        static constexpr int TOTAL_MOVES = BOARD_SIZE * BOARD_SIZE + 1;
        static constexpr int START_X = 100;
        static constexpr int START_Y = 30;
        static constexpr int CELL_SIZE = 55;
        static constexpr int CELL_DRAW_SIZE = 50;
        static constexpr int KNIGHT_SIZE = 35;

        std::vector<std::vector<int>> board;
        std::vector<Position> moveSequence;
        Position knightPos;
        int moves;
        bool tourOver;

        static constexpr std::array<int, 8> horizontal = {2, 1, -1, -2, -2, -1, 1, 2};
        static constexpr std::array<int, 8> vertical = {-1, -2, -2, -1, 1, 2, 2, 1};
    };

    class KnightsTourWindow : public mxvk::VK_Window {
      public:
        KnightsTourWindow(const std::string &path, int width, int height, bool fullscreen, bool enableVsync)
            : mxvk::VK_Window("Knights Tour", width, height, fullscreen, MXVK_VALIDATION, enableVsync),
              assetRoot((path.empty() || path == ".") ? std::string(KNIGHT_ASSET_DIR) : path),
              fontPath(assetRoot + "/data/font.ttf"),
              introStarted(SDL_GetTicks()) {
            setClearColor(0.0f, 0.0f, 0.0f, 1.0f);
            setFont(fontPath, TEXT_SIZE);
            loadWindowIcon();

            intro = createSprite(assetRoot + "/data/logo.png", "", assetRoot + "/data/fade.frag.spv");
            whiteCell = makeSolidSprite(255, 255, 255, 255);
            redCell = makeSolidSprite(255, 0, 0, 255);
            visitedCell = makeSolidSprite(0, 0, 0, 255);
            knightSprite = createSprite(assetRoot + "/data/knight.png", "", assetRoot + "/data/color_key.frag.spv");

            if (mxvk::Joystick::joysticks() > 0) {
                if (joystick.open(0)) {
                    std::cout << std::format("Joystick opened: {}\n", joystick.name());
                } else {
                    std::cout << "Could not open joystick..\n";
                }
            }
        }

        void event(SDL_Event &event) override {
            if (event.type == SDL_EVENT_KEY_DOWN) {
                if (event.key.key == SDLK_ESCAPE) {
                    exit();
                    return;
                }
                if (event.key.key == SDLK_S && !event.key.repeat) {
                    try {
                        saveSnapshot("screenshot.png");
                        std::cout << "mx: Screenshot captured..\n";
                    } catch (const mxvk::Exception &exception) {
                        std::cerr << std::format("mxvk: screenshot failed: {}\n", exception.text());
                    }
                    return;
                }
                if (screen != Screen::Tour) {
                    return;
                }
                if (event.key.key == SDLK_SPACE) {
                    tour.nextMove();
                } else if (event.key.key == SDLK_RETURN && !event.key.repeat) {
                    tour.resetTour();
                }
                return;
            }

            if (screen != Screen::Tour) {
                return;
            }

            if (event.type == SDL_EVENT_MOUSE_BUTTON_DOWN) {
                if (event.button.button == SDL_BUTTON_LEFT) {
                    resetTourFromMousePosition(event.button.x, event.button.y);
                } else if (event.button.button == SDL_BUTTON_RIGHT) {
                    tour.resetTour();
                }
            } else if (event.type == SDL_EVENT_JOYSTICK_BUTTON_DOWN) {
                if (event.jbutton.button == 1) {
                    tour.nextMove();
                } else if (event.jbutton.button == 2) {
                    tour.resetTour();
                }
            }
        }

        void proc() override {
            const float scaleX = swapchain_extent.width > 0U ? static_cast<float>(swapchain_extent.width) / DESIGN_WIDTH : 1.0f;
            const float scaleY = swapchain_extent.height > 0U ? static_cast<float>(swapchain_extent.height) / DESIGN_HEIGHT : 1.0f;

            if (screen == Screen::Intro) {
                drawIntro();
                return;
            }

            updateFont(scaleY);
            tour.drawBoard(*whiteCell, *redCell, *visitedCell, scaleX, scaleY);
            tour.drawKnight(*knightSprite, scaleX, scaleY);

            if (!tour.isTourOver()) {
                printScaledText("Knights Tour - Space to Move, Click a Square to Restart", TEXT_OFFSET_X, TEXT_OFFSET_Y, scaleX, scaleY);
                printMoveCount(scaleX, scaleY);
            } else {
                printScaledText("-[ Tour Complete ]- Press Return to Reset", TEXT_OFFSET_X, TEXT_OFFSET_Y, scaleX, scaleY);
            }
        }

      private:
        enum class Screen {
            Intro,
            Tour
        };

        static constexpr float DESIGN_WIDTH = 640.0f;
        static constexpr float DESIGN_HEIGHT = 480.0f;
        static constexpr int TEXT_OFFSET_X = 15;
        static constexpr int TEXT_OFFSET_Y = 5;
        static constexpr int TEXT_SIZE = 14;
        static constexpr Uint64 INTRO_STEP_MS = 15;
        static constexpr int INTRO_ALPHA_STEP = 3;

        std::string assetRoot;
        std::string fontPath;
        mxvk::VK_Sprite *intro = nullptr;
        mxvk::VK_Sprite *knightSprite = nullptr;
        mxvk::VK_Sprite *whiteCell = nullptr;
        mxvk::VK_Sprite *redCell = nullptr;
        mxvk::VK_Sprite *visitedCell = nullptr;
        mxvk::Joystick joystick;
        Tour tour;
        Screen screen = Screen::Intro;
        Uint64 introStarted = 0;
        int currentFontSize = TEXT_SIZE;

        mxvk::VK_Sprite *makeSolidSprite(std::uint8_t red, std::uint8_t green, std::uint8_t blue, std::uint8_t alpha) {
            const std::array<std::uint8_t, 4> pixel = {red, green, blue, alpha};
            mxvk::VK_Sprite *sprite = createSprite(1, 1);
            sprite->updateTexture(pixel.data(), 1, 1, 4);
            return sprite;
        }

        void loadWindowIcon() {
            const char *videoDriver = SDL_GetCurrentVideoDriver();
            if (videoDriver != nullptr && std::string(videoDriver) == "wayland") {
                return;
            }

            std::unique_ptr<SDL_Surface, decltype(&SDL_DestroySurface)> icon(
                mxvk::LoadPNG((assetRoot + "/data/knight.png").c_str()), SDL_DestroySurface);
            if (icon != nullptr && !SDL_SetWindowIcon(getSDLWindow(), icon.get())) {
                std::cerr << std::format("knight: could not set window icon: {}\n", SDL_GetError());
            }
        }

        void drawIntro() {
            const Uint64 elapsed = SDL_GetTicks() - introStarted;
            const int fadeSteps = static_cast<int>(elapsed / INTRO_STEP_MS);
            const int alpha = std::max(0, 255 - fadeSteps * INTRO_ALPHA_STEP);
            if (alpha == 0) {
                screen = Screen::Tour;
                return;
            }

            intro->setShaderParams(static_cast<float>(alpha) / 255.0f);
            intro->drawSpriteRect(0, 0, static_cast<int>(swapchain_extent.width), static_cast<int>(swapchain_extent.height));
        }

        void updateFont(float scaleY) {
            const int desiredSize = std::max(1, static_cast<int>(std::lround(TEXT_SIZE * scaleY)));
            if (desiredSize == currentFontSize) {
                return;
            }
            setFont(fontPath, desiredSize);
            currentFontSize = desiredSize;
        }

        void printScaledText(const std::string &text, int x, int y, float scaleX, float scaleY) {
            printText(text,
                      static_cast<int>(std::lround(x * scaleX)),
                      static_cast<int>(std::lround(y * scaleY)),
                      SDL_Color{255, 255, 255, 255});
        }

        void printMoveCount(float scaleX, float scaleY) {
            const std::string text = std::format("Moves: {}", tour.getMoves());
            int textWidth = 0;
            int textHeight = 0;
            const int rightMargin = static_cast<int>(std::lround(TEXT_OFFSET_X * scaleX));
            int x = static_cast<int>(std::lround(400.0f * scaleX));
            if (getTextDimensions(text, textWidth, textHeight)) {
                x = std::max(0, static_cast<int>(swapchain_extent.width) - textWidth - rightMargin);
            }
            printText(text,
                      x,
                      static_cast<int>(std::lround(TEXT_OFFSET_Y * scaleY)),
                      SDL_Color{255, 255, 255, 255});
        }

        void resetTourFromMousePosition(float mouseX, float mouseY) {
            int windowWidth = 0;
            int windowHeight = 0;
            SDL_GetWindowSize(getSDLWindow(), &windowWidth, &windowHeight);
            if (windowWidth <= 0 || windowHeight <= 0) {
                return;
            }

            const float designX = mouseX * DESIGN_WIDTH / static_cast<float>(windowWidth);
            const float designY = mouseY * DESIGN_HEIGHT / static_cast<float>(windowHeight);
            tour.resetTourFromPoint(designX, designY);
        }
    };

    Tour::Tour() : moves(1), tourOver(false) {
        std::srand(static_cast<unsigned int>(std::time(nullptr)));
        initializeBoard();
        resetTour();
    }

    void Tour::initializeBoard() {
        board.resize(BOARD_SIZE, std::vector<int>(BOARD_SIZE, 0));
    }

    void Tour::clearBoard() {
        for (auto &row : board) {
            std::fill(row.begin(), row.end(), 0);
        }
    }

    bool Tour::isValidMove(const Position &position) const {
        return position.row >= 0 && position.row < BOARD_SIZE &&
               position.col >= 0 && position.col < BOARD_SIZE &&
               board[position.row][position.col] == 0;
    }

    int Tour::getDegree(const Position &position) const {
        int count = 0;
        for (int index = 0; index < 8; ++index) {
            const int newRow = position.row + vertical[index];
            const int newCol = position.col + horizontal[index];
            if (isValidMove(Position(newRow, newCol))) {
                ++count;
            }
        }
        return count;
    }

    bool Tour::solveKnightsTour(Position position, int moveCount) {
        if (moveCount == TOTAL_MOVES) {
            return true;
        }

        std::vector<std::pair<int, Position>> nextMoves;
        for (int index = 0; index < 8; ++index) {
            Position nextPosition(position.row + vertical[index], position.col + horizontal[index]);
            if (isValidMove(nextPosition)) {
                nextMoves.emplace_back(getDegree(nextPosition), nextPosition);
            }
        }

        std::sort(nextMoves.begin(), nextMoves.end(), [](const auto &left, const auto &right) {
            return left.first < right.first;
        });

        for (const auto &[degree, nextPosition] : nextMoves) {
            [[maybe_unused]] const int moveDegree = degree;
            board[nextPosition.row][nextPosition.col] = moveCount;
            moveSequence.push_back(nextPosition);

            if (solveKnightsTour(nextPosition, moveCount + 1)) {
                return true;
            }

            board[nextPosition.row][nextPosition.col] = 0;
            moveSequence.pop_back();
        }
        return false;
    }

    void Tour::resetTour() {
        resetTour(std::rand() % BOARD_SIZE, std::rand() % BOARD_SIZE);
    }

    void Tour::resetTour(int startRow, int startCol) {
        if (startRow < 0 || startRow >= BOARD_SIZE || startCol < 0 || startCol >= BOARD_SIZE) {
            return;
        }

        clearBoard();
        knightPos = Position(startRow, startCol);
        board[knightPos.row][knightPos.col] = 1;
        moveSequence.clear();
        moveSequence.push_back(knightPos);
        solveKnightsTour(knightPos, 2);
        moves = 1;
        tourOver = false;
    }

    void Tour::resetTourFromPoint(float x, float y) {
        const int localX = static_cast<int>(std::floor(x)) - START_X;
        const int localY = static_cast<int>(std::floor(y)) - START_Y;
        if (localX < 0 || localY < 0) {
            return;
        }

        const int col = localX / CELL_SIZE;
        const int row = localY / CELL_SIZE;
        if (row >= BOARD_SIZE || col >= BOARD_SIZE ||
            localX % CELL_SIZE >= CELL_DRAW_SIZE || localY % CELL_SIZE >= CELL_DRAW_SIZE) {
            return;
        }

        resetTour(row, col);
    }

    void Tour::nextMove() {
        if (tourOver || static_cast<std::size_t>(moves) >= moveSequence.size()) {
            return;
        }

        const Position nextPosition = moveSequence[static_cast<std::size_t>(moves)];
        board[knightPos.row][knightPos.col] = -1;
        knightPos = nextPosition;
        ++moves;
        board[knightPos.row][knightPos.col] = moves;
        tourOver = static_cast<std::size_t>(moves) == moveSequence.size();
    }

    void Tour::drawBoard(mxvk::VK_Sprite &whiteCell, mxvk::VK_Sprite &redCell, mxvk::VK_Sprite &visitedCell, float scaleX, float scaleY) const {
        for (int row = 0; row < BOARD_SIZE; ++row) {
            for (int col = 0; col < BOARD_SIZE; ++col) {
                mxvk::VK_Sprite *cell = nullptr;
                if (board[row][col] == -1) {
                    cell = &visitedCell;
                } else if ((row + col) % 2 == 0) {
                    cell = &whiteCell;
                } else {
                    cell = &redCell;
                }

                cell->drawSpriteRect(
                    static_cast<int>(std::lround((START_X + col * CELL_SIZE) * scaleX)),
                    static_cast<int>(std::lround((START_Y + row * CELL_SIZE) * scaleY)),
                    static_cast<int>(std::lround(CELL_DRAW_SIZE * scaleX)),
                    static_cast<int>(std::lround(CELL_DRAW_SIZE * scaleY)));
            }
        }
    }

    void Tour::drawKnight(mxvk::VK_Sprite &texture, float scaleX, float scaleY) const {
        texture.drawSpriteRect(
            static_cast<int>(std::lround((START_X + knightPos.col * CELL_SIZE + 5) * scaleX)),
            static_cast<int>(std::lround((START_Y + knightPos.row * CELL_SIZE + 5) * scaleY)),
            static_cast<int>(std::lround(KNIGHT_SIZE * scaleX)),
            static_cast<int>(std::lround(KNIGHT_SIZE * scaleY)));
    }
} // namespace knight

int main(int argc, char **argv) {
    try {
        Arguments args = proc_args(argc, argv);
        if (!args.resolutionSpecified) {
            args.width = 960;
            args.height = 720;
        }
        knight::KnightsTourWindow window(args.path, args.width, args.height, args.fullscreen, args.enable_vsync);
        window.loop();
    } catch (mxvk::Exception &exception) {
        std::cerr << std::format("mxvk: Exception: {}\n", exception.text());
        return EXIT_FAILURE;
    } catch (ArgException<std::string> &exception) {
        std::cerr << std::format("mxvk: Argument Exception: {}\n", exception.text());
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

File: python-examples/knight/knight.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import math
import random
import sys
import time
from dataclasses import dataclass
from pathlib import Path

import numpy as np

import mxvk_ext as mxvk


EXAMPLE_DIR = Path(__file__).resolve().parent


def lround(value: float) -> int:
    if (value >= 0.0):
        return math.floor(value + 0.5)

    return math.ceil(value - 0.5)


@dataclass(frozen=True, slots=True)
class Position:
    row: int = 0
    col: int = 0


class Tour:
    BOARD_SIZE = 8
    TOTAL_MOVES = BOARD_SIZE * BOARD_SIZE + 1

    START_X = 100
    START_Y = 30

    CELL_SIZE = 55
    CELL_DRAW_SIZE = 50
    KNIGHT_SIZE = 35

    HORIZONTAL = (2, 1, -1, -2, -2, -1, 1, 2)
    VERTICAL = (-1, -2, -2, -1, 1, 2, 2, 1)

    def __init__(self):
        self.board = [[0 for _ in range(self.BOARD_SIZE)] for _ in range(self.BOARD_SIZE)]
        self.move_sequence: list[Position] = []
        self.knight_pos = Position()
        self.moves = 1
        self.tour_over = False
        self.reset_tour()

    def clear_board(self):
        for row in range(self.BOARD_SIZE):
            for col in range(self.BOARD_SIZE):
                self.board[row][col] = 0

    def is_valid_move(self, position: Position) -> bool:
        return (
            0 <= position.row < self.BOARD_SIZE
            and 0 <= position.col < self.BOARD_SIZE
            and self.board[position.row][position.col] == 0
        )

    def get_degree(self, position: Position) -> int:
        count = 0

        for index in range(8):
            new_row = position.row + self.VERTICAL[index]
            new_col = position.col + self.HORIZONTAL[index]

            if (self.is_valid_move(Position(new_row, new_col))):
                count += 1

        return count

    def solve_knights_tour(self, position: Position, move_count: int) -> bool:
        if (move_count == self.TOTAL_MOVES):
            return True

        next_moves: list[tuple[int, Position]] = []

        for index in range(8):
            next_position = Position(position.row + self.VERTICAL[index], position.col + self.HORIZONTAL[index])

            if (self.is_valid_move(next_position)):
                next_moves.append((self.get_degree(next_position), next_position))

        next_moves.sort(key=lambda item: item[0])

        for _, next_position in next_moves:
            self.board[next_position.row][next_position.col] = move_count
            self.move_sequence.append(next_position)

            if (self.solve_knights_tour(next_position, move_count + 1)):
                return True

            self.board[next_position.row][next_position.col] = 0
            self.move_sequence.pop()

        return False

    def reset_tour(self, start_row: int | None = None, start_col: int | None = None):
        if (start_row is None or start_col is None):
            start_row = random.randrange(self.BOARD_SIZE)
            start_col = random.randrange(self.BOARD_SIZE)

        if (start_row < 0 or start_row >= self.BOARD_SIZE or start_col < 0 or start_col >= self.BOARD_SIZE):
            return

        self.clear_board()

        self.knight_pos = Position(start_row, start_col)
        self.board[self.knight_pos.row][self.knight_pos.col] = 1

        self.move_sequence.clear()
        self.move_sequence.append(self.knight_pos)

        self.solve_knights_tour(self.knight_pos, 2)

        self.moves = 1
        self.tour_over = False

    def reset_tour_from_point(self, x: float, y: float):
        local_x = math.floor(x) - self.START_X
        local_y = math.floor(y) - self.START_Y

        if (local_x < 0 or local_y < 0):
            return

        col = local_x // self.CELL_SIZE
        row = local_y // self.CELL_SIZE

        if (row >= self.BOARD_SIZE or col >= self.BOARD_SIZE or local_x % self.CELL_SIZE >= self.CELL_DRAW_SIZE or local_y % self.CELL_SIZE >= self.CELL_DRAW_SIZE):
            return

        self.reset_tour(row, col)

    def next_move(self):
        if (self.tour_over or self.moves >= len(self.move_sequence)):
            return

        next_position = self.move_sequence[self.moves]

        self.board[self.knight_pos.row][self.knight_pos.col] = -1
        self.knight_pos = next_position
        self.moves += 1
        self.board[self.knight_pos.row][self.knight_pos.col] = self.moves
        self.tour_over = self.moves == len(self.move_sequence)

    def get_moves(self) -> int:
        return self.moves

    def is_tour_over(self) -> bool:
        return self.tour_over

    def draw_board(self, white_cell, red_cell, visited_cell, scale_x: float, scale_y: float):
        for row in range(self.BOARD_SIZE):
            for col in range(self.BOARD_SIZE):
                if (self.board[row][col] == -1):
                    cell = visited_cell
                elif ((row + col) % 2 == 0):
                    cell = white_cell
                else:
                    cell = red_cell

                x = lround((self.START_X + col * self.CELL_SIZE) * scale_x)
                y = lround((self.START_Y + row * self.CELL_SIZE) * scale_y)
                width = lround(self.CELL_DRAW_SIZE * scale_x)
                height = lround(self.CELL_DRAW_SIZE * scale_y)

                cell.draw_rect(x, y, width, height)

    def draw_knight(self, texture, scale_x: float, scale_y: float):
        x = lround((self.START_X + self.knight_pos.col * self.CELL_SIZE + 5) * scale_x)
        y = lround((self.START_Y + self.knight_pos.row * self.CELL_SIZE + 5) * scale_y)
        width = lround(self.KNIGHT_SIZE * scale_x)
        height = lround(self.KNIGHT_SIZE * scale_y)

        texture.draw_rect(x, y, width, height)


class KnightsTourWindow(mxvk.VK_Window):
    DESIGN_WIDTH = 640.0
    DESIGN_HEIGHT = 480.0

    TEXT_OFFSET_X = 15
    TEXT_OFFSET_Y = 5
    TEXT_SIZE = 14

    INTRO_STEP_MS = 15
    INTRO_ALPHA_STEP = 3

    SCREEN_INTRO = 0
    SCREEN_TOUR = 1

    KEY_S = mxvk.key_code("S")
    KEY_SPACE = mxvk.key_code("Space")
    KEY_RETURN = mxvk.key_code("Return")

    def __init__(self, path: str, width: int, height: int, fullscreen: bool, enable_vsync: bool):
        self.asset_root = Path(path).resolve() if path else EXAMPLE_DIR
        self.data_directory = self.asset_root / "data"
        mxvk.set_default_shader_directory(str(self.data_directory))

        super().__init__("Knights Tour", width, height, fullscreen, False, enable_vsync)

        self.font_path = str(self.data_directory / "font.ttf")
        self.screen = self.SCREEN_INTRO
        self.intro_started = time.monotonic()
        self.current_font_size = self.TEXT_SIZE
        self.text_color = mxvk.Color(255, 255, 255, 255)

        self.tour = Tour()

        self.set_clear_color(0.0, 0.0, 0.0, 1.0)
        self.set_font(self.font_path, self.TEXT_SIZE)

        sprite_vertex_shader = str(self.data_directory / "sprite.vert.spv")
        self.intro = self.create_sprite(str(self.data_directory / "logo.png"), sprite_vertex_shader, str(self.data_directory / "fade.frag.spv"))

        self.white_cell = self.make_solid_sprite(255, 255, 255, 255)
        self.red_cell = self.make_solid_sprite(255, 0, 0, 255)
        self.visited_cell = self.make_solid_sprite(0, 0, 0, 255)

        self.knight_sprite = self.create_sprite(str(self.data_directory / "knight.png"), sprite_vertex_shader, str(self.data_directory / "color_key.frag.spv"))

        self.joystick = mxvk.Joystick()
        self.joystick_open = False
        self.joystick_button_count = 0

        self.previous_button_1 = False
        self.previous_button_2 = False

        if (mxvk.Joystick.count() > 0):
            if (self.joystick.open(0)):
                self.joystick_open = True
                self.joystick_button_count = self.joystick.num_buttons()
                print(f"Joystick opened: {self.joystick.name()}")
            else:
                print("Could not open joystick..")

    def close(self):
        if self.joystick_open:
            self.joystick.close()
            self.joystick_open = False
        self.intro = None
        self.white_cell = None
        self.red_cell = None
        self.visited_cell = None
        self.knight_sprite = None
        self.release()

    def make_solid_sprite(self, red: int, green: int, blue: int, alpha: int):
        pixel = np.array([[[red, green, blue, alpha]]], dtype=np.uint8)

        sprite = self.create_sprite(1, 1)
        sprite.update_texture(pixel, 1, 1, 4)

        return sprite

    def event(self, event):
        if (event.key_down):
            if (event.key == mxvk.KEY_ESCAPE):
                self.request_exit()
                return

            if (event.key == self.KEY_S and not event.repeat):
                try:
                    self.save_snapshot("screenshot.png")
                    print("mx: Screenshot captured..")

                except mxvk.MXVKError as exception:
                    print(f"mxvk: screenshot failed: {exception}", file=sys.stderr)

                return

            if (self.screen != self.SCREEN_TOUR):
                return

            if (event.key == self.KEY_SPACE):
                self.tour.next_move()

            elif (event.key == self.KEY_RETURN and not event.repeat):
                self.tour.reset_tour()

            return

        if (self.screen != self.SCREEN_TOUR):
            return

        if (event.mouse_button_down):
            if (event.button == mxvk.MOUSE_BUTTON_LEFT):
                self.reset_tour_from_mouse_position(event.x, event.y)

            elif (event.button == mxvk.MOUSE_BUTTON_RIGHT):
                self.tour.reset_tour()

    def poll_joystick(self):
        if (not self.joystick_open):
            return

        button_1 = False
        button_2 = False

        if (self.joystick_button_count > 1):
            button_1 = bool(self.joystick.button(1))

        if (self.joystick_button_count > 2):
            button_2 = bool(self.joystick.button(2))

        if (self.screen == self.SCREEN_TOUR):
            if (button_1 and not self.previous_button_1):
                self.tour.next_move()

            if (button_2 and not self.previous_button_2):
                self.tour.reset_tour()

        self.previous_button_1 = button_1
        self.previous_button_2 = button_2

    def proc(self):
        self.poll_joystick()

        swap_width, swap_height = self.swapchain_extent

        if (swap_width > 0):
            scale_x = float(swap_width) / self.DESIGN_WIDTH
        else:
            scale_x = 1.0

        if (swap_height > 0):
            scale_y = float(swap_height) / self.DESIGN_HEIGHT
        else:
            scale_y = 1.0

        if (self.screen == self.SCREEN_INTRO):
            self.draw_intro()
            return

        self.update_font(scale_y)

        self.tour.draw_board(self.white_cell, self.red_cell, self.visited_cell, scale_x, scale_y)
        self.tour.draw_knight(self.knight_sprite, scale_x, scale_y)

        if (not self.tour.is_tour_over()):
            self.print_scaled_text("Knights Tour - Space to Move, Click a Square to Restart", self.TEXT_OFFSET_X, self.TEXT_OFFSET_Y, scale_x, scale_y)
            self.print_move_count(scale_x, scale_y)
        else:
            self.print_scaled_text("-[ Tour Complete ]- Press Return to Reset", self.TEXT_OFFSET_X, self.TEXT_OFFSET_Y, scale_x, scale_y)

    def draw_intro(self):
        elapsed_ms = int((time.monotonic() - self.intro_started) * 1000.0)
        fade_steps = elapsed_ms // self.INTRO_STEP_MS
        alpha = max(0, 255 - fade_steps * self.INTRO_ALPHA_STEP)

        if (alpha == 0):
            self.screen = self.SCREEN_TOUR
            return

        swap_width, swap_height = self.swapchain_extent

        self.intro.set_shader_params(float(alpha) / 255.0)
        self.intro.draw_rect(0, 0, int(swap_width), int(swap_height))

    def update_font(self, scale_y: float):
        desired_size = max(1, lround(self.TEXT_SIZE * scale_y))

        if (desired_size == self.current_font_size):
            return

        self.set_font(self.font_path, desired_size)
        self.current_font_size = desired_size

    def print_scaled_text(self, text: str, x: int, y: int, scale_x: float, scale_y: float):
        self.print_text(text, lround(x * scale_x), lround(y * scale_y), self.text_color)

    def print_move_count(self, scale_x: float, scale_y: float):
        text = f"Moves: {self.tour.get_moves()}"
        swap_width, _ = self.swapchain_extent
        right_margin = lround(self.TEXT_OFFSET_X * scale_x)
        x = lround(400.0 * scale_x)

        dimensions = self.get_text_dimensions(text)

        if (dimensions is not None):
            text_width, _ = dimensions
            x = max(0, int(swap_width) - int(text_width) - right_margin)

        self.print_text(text, x, lround(self.TEXT_OFFSET_Y * scale_y), self.text_color)

    def reset_tour_from_mouse_position(self, mouse_x: float, mouse_y: float):
        window_width, window_height = self.swapchain_extent

        if (window_width <= 0 or window_height <= 0):
            return

        design_x = mouse_x * self.DESIGN_WIDTH / float(window_width)
        design_y = mouse_y * self.DESIGN_HEIGHT / float(window_height)

        self.tour.reset_tour_from_point(design_x, design_y)


def parse_resolution(value: str) -> tuple[int, int]:
    try:
        width_text, height_text = value.lower().split("x", 1)
        width = int(width_text)
        height = int(height_text)

    except ValueError as exception:
        raise argparse.ArgumentTypeError("resolution must use WIDTHxHEIGHT, for example 960x720") from exception

    if (width <= 0 or height <= 0):
        raise argparse.ArgumentTypeError("resolution width and height must be greater than zero")

    return width, height


def parse_args():
    parser = argparse.ArgumentParser(description="Knights Tour using the MXVK Python bindings")

    parser.add_argument("-p", "--path", default=str(EXAMPLE_DIR), help="resource root containing the data directory")
    parser.add_argument("-r", "--resolution", type=parse_resolution, default=(960, 720), metavar="WIDTHxHEIGHT", help="window resolution (default: 960x720)")
    parser.add_argument("-f", "--fullscreen", action="store_true", help="start fullscreen")
    parser.add_argument("--enable-vsync", action="store_true", help="enable vertical synchronization")

    return parser.parse_args()


def main() -> int:
    args = parse_args()
    width, height = args.resolution
    window = None

    try:
        window = KnightsTourWindow(args.path, width, height, args.fullscreen, args.enable_vsync)
        window.loop()

    except mxvk.MXVKError as exception:
        print(f"mxvk: Exception: {exception}", file=sys.stderr)
        return 1

    finally:
        if window is not None:
            window.close()

    return 0


if (__name__ == "__main__"):
    raise SystemExit(main())
